### ISSUE TITLE
Prop `aria-controls` did not match.

### ISSUE THREAD

**Issue by EmircanDuman** [shadcn-ui/ui#1018]
I installed, popover, command and button with npx and copied over the first combobox component from the docs but I am getting the following error

`Warning: Prop aria-controls did not match. Server: "radix-:R1mcq:" Client: "radix-:R6pj9:"
    at button
    at _c (webpack-internal:///(app-pages-browser)/./components/ui/button.tsx:40:11)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-slot/dist/index.mjs:46:23)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-slot/dist/index.mjs:20:23)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-primitive/dist/index.mjs:44:26)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-slot/dist/index.mjs:46:23)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-slot/dist/index.mjs:20:23)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-primitive/dist/index.mjs:44:26)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-popper/dist/index.mjs:80:28)
    at eval (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-popover/dist/index.mjs:139:29)
    at Provider (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-context/dist/index.mjs:47:28)
    at Provider (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-context/dist/index.mjs:47:28)
    at $cf1ac5d9fe0e8206$export$badac9ada3a0bdf9 (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-popper/dist/index.mjs:65:28)
    at $cb5cc270b50c6fcd$export$5b6b19405a83ff9d (webpack-internal:///(app-pages-browser)/./node_modules/@radix-ui/react-popover/dist/index.mjs:81:29)
    at CoinSelector (webpack-internal:///(app-pages-browser)/./components/CoinSelector.tsx:47:64)
    at main
    at InnerLayoutRouter (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/layout-router.js:228:11)
    at RedirectErrorBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/redirect-boundary.js:73:9)
    at RedirectBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/redirect-boundary.js:81:11)
    at NotFoundBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/not-found-boundary.js:62:11)
    at LoadingBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/layout-router.js:325:11)
    at ErrorBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/error-boundary.js:104:11)
    at InnerScrollAndFocusHandler (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/layout-router.js:139:9)
    at ScrollAndFocusHandler (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/layout-router.js:214:11)
    at RenderFromTemplateContext (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/render-from-template-context.js:15:44)
    at OuterLayoutRouter (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/layout-router.js:335:11)
    at body
    at html
    at RedirectErrorBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/redirect-boundary.js:73:9)
    at RedirectBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/redirect-boundary.js:81:11)
    at NotFoundErrorBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/not-found-boundary.js:54:9)
    at NotFoundBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/not-found-boundary.js:62:11)
    at DevRootNotFoundBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/dev-root-not-found-boundary.js:32:11)
    at ReactDevOverlay (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/react-dev-overlay/internal/ReactDevOverlay.js:66:9)
    at HotReload (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/react-dev-overlay/hot-reloader-client.js:326:11)
    at Router (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/app-router.js:148:11)
    at ErrorBoundaryHandler (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/error-boundary.js:77:9)
    at ErrorBoundary (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/error-boundary.js:104:11)
    at AppRouter (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/app-router.js:396:13)
    at ServerRoot (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/app-index.js:166:11)
    at RSCComponent
    at ReactDevOverlay (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/components/react-dev-overlay/internal/ReactDevOverlay.js:66:9)
    at Root (webpack-internal:///(app-pages-browser)/./node_modules/next/dist/client/app-index.js:183:11)`

I am new to this library and I cannot pinpoint the problem. There are no adblocker extensions on my web browser

**Comment by Saim1358** [comment]
I am also facing the same error when i am trying to use the tabs component. The component is working fine but when i open the browser console it shows this error . It is quite annoying. And i think there is some error in the core of the component or something related to radix ui . ☹️

**Comment by asheikho99** [comment]
I have the same issue but while using React Hook Form component example from the docs without anything else in the page.

![image](https://github.com/shadcn-ui/ui/assets/105573050/d3105400-e780-4579-a9ca-36eaa6e410d6)

**Comment by nikolubbe** [comment]
Looks like a NextJS bug: https://github.com/vercel/next.js/issues/53110

**Comment by Saim1358** [comment]
It should be fixed as soon as possible 🙂

**Comment by daniel-koti** [comment]
> I am also facing the same error when i am trying to use the tabs component. The component is working fine but when i open the browser console it shows this error . It is quite annoying. And i think there is some error in the core of the component or something related to radix ui . ☹️

I have the same problem ☹️

**Comment by she0305** [comment]
Can confirm the bug here as well :bug:

**Comment by emirefek** [comment]
Having same problem. Having same thing on Select and Dropdown components for now. Probably will have more while deving app. It's my first try with Shadcn/ui. 

Any way to fix?

"next": "13.4.12"

**Comment by DomNidy** [comment]
For anyone having this issue, I was able to suppress the error by giving my `<SelectTrigger>` component the `aria-controls` prop of "radix-:R1mcq:" (or whatever the class was on the server side in your error)

My code for reference
```
<Select onValueChange={(v) => console.log(v)}>
        <SelectTrigger className="w-[180px]" aria-controls="radix-:R1mcq:">
          <SelectValue placeholder="Theme"></SelectValue>
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="light">Light</SelectItem>
          <SelectItem value="dark">Dark</SelectItem>
          <SelectItem value="system">System</SelectItem>
        </SelectContent>
</Select>

```

**Comment by daniel-koti** [comment]
> For anyone having this issue, I was able to suppress the error by giving my `<SelectTrigger>` component the `aria-controls` prop of "radix-:R1mcq:" (or whatever the class was on the server side in your error)
> 
> Assuming this should work for other shadcn components as well
> 
> My code for reference
> 
> ```
> <Select onValueChange={(v) => console.log(v)}>
>         <SelectTrigger className="w-[180px]" aria-controls="radix-:R1mcq:">
>           <SelectValue placeholder="Theme"></SelectValue>
>         </SelectTrigger>
>         <SelectContent>
>           <SelectItem value="light">Light</SelectItem>
>           <SelectItem value="dark">Dark</SelectItem>
>           <SelectItem value="system">System</SelectItem>
>         </SelectContent>
> </Select>
> ```


I'm using Tabs component, and it didn't work for me 🥲.

**Comment by EmircanDuman** [comment]
> For anyone having this issue, I was able to suppress the error by giving my `<SelectTrigger>` component the `aria-controls` prop of "radix-:R1mcq:" (or whatever the class was on the server side in your error)
> 
> Assuming this should work for other shadcn components as well
> 
> My code for reference
> 
> ```
> <Select onValueChange={(v) => console.log(v)}>
>         <SelectTrigger className="w-[180px]" aria-controls="radix-:R1mcq:">
>           <SelectValue placeholder="Theme"></SelectValue>
>         </SelectTrigger>
>         <SelectContent>
>           <SelectItem value="light">Light</SelectItem>
>           <SelectItem value="dark">Dark</SelectItem>
>           <SelectItem value="system">System</SelectItem>
>         </SelectContent>
> </Select>
> ```

This solution worked for me, thanks a lot!

**Comment by aeholic** [comment]
Same problem here but it only happens on Firefox for me. Chrome and Safari is all fine and I am building a form with these components.

**Comment by 3dln** [comment]
> Same problem here but it only happens on Firefox for me. Chrome and Safari is all fine and I am building a form with these components.

You can try install the canary version `next@canary`

**Comment by picklebrownie** [comment]
This worked for me 

```
<PopoverTrigger asChild>
        <Button
          variant="outline"
          size="sm"
          role="combobox"
          aria-expanded={open}
          aria-label="Select a store"
          aria-controls="radix-:R1arbdj9:"
          className={cn('w-[200px] justify-between', className)}
        >
          <StoreIcon className="mr-2 h-4 w-4" />
          {currentStore?.label}
          <ChevronsDownUp className="ml-auto h-4 w-4 shrink-0 opacity-50" />
        </Button>
 </PopoverTrigger>
```

Thank You

**Comment by nmajor** [comment]
Looks like this is an issue with nextjs: https://github.com/radix-ui/primitives/issues/1684

**Comment by jjalonso** [comment]
to me combobox exampe from the web give me this error on Button 
`"Types of property ''aria-expanded'' are incompatible."`

**Comment by adham618** [comment]
> For anyone having this issue, I was able to suppress the error by giving my `<SelectTrigger>` component the `aria-controls` prop of "radix-:R1mcq:" (or whatever the class was on the server side in your error)
> 
> Assuming this should work for other shadcn components as well
> 
> My code for reference
> 
> ```
> <Select onValueChange={(v) => console.log(v)}>
>         <SelectTrigger className="w-[180px]" aria-controls="radix-:R1mcq:">
>           <SelectValue placeholder="Theme"></SelectValue>
>         </SelectTrigger>
>         <SelectContent>
>           <SelectItem value="light">Light</SelectItem>
>           <SelectItem value="dark">Dark</SelectItem>
>           <SelectItem value="system">System</SelectItem>
>         </SelectContent>
> </Select>
> ```

for the accordion or tabs, you also have to add id or aria-labelledb for the content if asked

**Comment by dushmanta05** [comment]
Faced the same issue

**Comment by joaoguidev** [comment]
This issue is still happening. I'm having this using the Tabs component on Remix.run 2.5.0

**Comment by dushmanta05** [comment]
@joaoguidev I was able to resolve the issue by keeping the server and client aria-controls values the same; you could change it in your code and it'll be resolved.

**Comment by gitMonks** [comment]
If you're encountering a hydration mismatch warning with Radix UI Tabs, it often stems from inconsistent ID generation between server and client rendering. Here's a streamlined approach to resolve this issue:
1. Ensure Client-Side Rendering: Use the 'use client' directive at the top of your component file to ensure it's rendered on the client side.
Use Stable IDs: Utilize the useId hook from React to generate stable IDs that are consistent across server and client renders.
3. Avoid Hardcoded IDs: Let Radix UI handle ID generation and ARIA attributes automatically.
Here's a minimal example:
```
'use client';

import { useEffect, useRef, useId } from 'react';
import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/app/components/ui/tabs';

const SimpleTabs = ({ tabs, activeTabId, onChangeActiveTab }) => {
  const stableId = useId();

  return (
    <Tabs value={activeTabId ?? ''} onValueChange={onChangeActiveTab} defaultValue={tabs[0]?.value}>
      <TabsList>
        {tabs.map((tab) => (
          <TabsTrigger key={`${stableId}-${tab.value}`} value={tab.value}>
            {tab.name}
          </TabsTrigger>
        ))}
      </TabsList>
      {tabs.map((tab) => (
        <TabsContent key={`${stableId}-content-${tab.value}`} value={tab.value}>
          <div>{tab.content}</div>
        </TabsContent>
      ))}
    </TabsUI>
  );
};

export default SimpleTabs;
```
Key Points:
Client-Side Directive: Ensure the component is client-rendered with 'use client'.
Stable ID Generation: Use useId to generate consistent IDs.
Radix UI Handling: Allow Radix UI to manage IDs and ARIA attributes to prevent mismatches.
By following these steps, you should be able to resolve the hydration mismatch warning effectively. If issues persist, ensure that all components using Radix UI Tabs are client-rendered and avoid any server-side logic that might interfere with ID generation.

**Comment by GermanGrib** [comment]
I used the <Dialog /> component from shadcn/ui and encountered a validation error in validator.w3, stating that `aria-controls` did not match. The issue was caused by the <Portal /> used inside this component.  

I tried everything, but none of the previously mentioned solutions worked for me.  

So, I decided to sacrifice screen readers and other assistive technologies by simply passing `undefined` to `aria-controls`, which removed the validation error. How this will affect SEO and other aspects, I’m not sure yet.
